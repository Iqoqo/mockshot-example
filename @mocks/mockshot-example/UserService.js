"use strict";
exports.__esModule = true;
var UserServiceMocks = /** @class */ (function () {
    function UserServiceMocks() {
    }
    UserServiceMocks.getUser = function (mock) {
        switch (mock) {
            case "success":
                return {
                    "email": "test@iqoqo.co",
                    "id": "abc",
                    "name": "Some Name"
                };
            default:
                throw Error("Unknown mock: " + mock);
        }
    };
    return UserServiceMocks;
}());
exports.UserServiceMocks = UserServiceMocks;
