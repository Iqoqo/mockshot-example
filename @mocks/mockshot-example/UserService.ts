export class UserServiceMocks {
    static getUser(mock: "success"): any {
        switch (mock) {
            case "success":
                return {
                  "email": "test@iqoqo.co",
                  "id": "abc",
                  "name": "Some Name"
                }
            default:
                throw Error("Unknown mock: "+mock);
        }
    }
}
