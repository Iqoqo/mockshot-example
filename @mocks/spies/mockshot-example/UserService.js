"use strict";
exports.__esModule = true;
var classTree = { "getUser": { "success": { "meta": { "key": "UserService Should get user: [mockshot] [ClassSnap] [[UserService getUser success]] 1", "originFile": "test/__snapshots__/UserService.spec.ts.snap" }, "mock": { "email": "test@iqoqo.co", "id": "abc", "name": "Some Name" } } } };
function getSpy(methodName) {
    var newSpy = jest.fn();
    var getMock = function (mockName) { return classTree[methodName][mockName].mock; };
    var mockImplementation = newSpy.mockImplementation, mockImplementationOnce = newSpy.mockImplementationOnce;
    newSpy.mockImplementation = function (mockName) {
        return typeof mockName === "string"
            ? mockImplementation(function () { return getMock(mockName); })
            : mockImplementation(mockName);
    };
    newSpy.mockImplementationOnce = function (mockName) {
        return typeof mockName === "string"
            ? mockImplementationOnce(function () { return getMock(mockName); })
            : mockImplementationOnce(mockName);
    };
    newSpy.getMock = getMock;
    return newSpy;
}
exports.userServiceSpy = {
    getUser: getSpy("getUser"),
};
