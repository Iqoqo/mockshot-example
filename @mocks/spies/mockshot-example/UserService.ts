interface MockshotMock<P, T = {}> extends jest.Mock {
    mockImplementation(mockName: P): jest.Mock<T>;
    mockImplementation(fn: (...args: any[]) => any): jest.Mock<T>;
    mockImplementationOnce(mockName: P): jest.Mock<T>;
    mockImplementationOnce(fn: (...args: any[]) => any): jest.Mock<T>;
    getMock(mockName: P): any;
}

interface UserServiceSpy {
    getUser: MockshotMock<"success">;
}

const classTree = {"getUser":{"success":{"meta":{"key":"UserService Should get user: [mockshot] [ClassSnap] [[UserService getUser success]] 1","originFile":"test/__snapshots__/UserService.spec.ts.snap"},"mock":{"email":"test@iqoqo.co","id":"abc","name":"Some Name"}}}};

function getSpy<P extends string>(methodName: string): MockshotMock<P> {
    const newSpy = jest.fn() as MockshotMock<P>;
    const getMock = mockName => classTree[methodName][mockName].mock;
    const { mockImplementation, mockImplementationOnce } = newSpy;
    newSpy.mockImplementation = mockName =>
      typeof mockName === "string"
        ? mockImplementation(() => getMock(mockName))
        : mockImplementation(mockName);
    newSpy.mockImplementationOnce = mockName =>
      typeof mockName === "string"
        ? mockImplementationOnce(() => getMock(mockName))
        : mockImplementationOnce(mockName);
    newSpy.getMock = getMock;
    return newSpy;
}

export const userServiceSpy: UserServiceSpy = {
    getUser: getSpy("getUser"),
}
;
