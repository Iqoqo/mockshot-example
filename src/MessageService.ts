
export class MessageService {

    constructor(private readonly userService){}

    async prepareMessage(text: string, userId: string) {
        const user = await this.userService.getUser(userId);
        return { 
            to: user.email,
            title: `Hello ${user.name}`,
            body: text
        }

    }
}