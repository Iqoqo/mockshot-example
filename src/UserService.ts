export interface IUser {
    id: string;
    name: string;
    email: string;
}

export class UserService {

    constructor(private readonly db){}

    async getUser(id: string):Promise<IUser>{
        const user = await this.db.findOneById(id);
        return {
            id: user.id,
            name: user.name, 
            email: user.email
        };
    }

}

