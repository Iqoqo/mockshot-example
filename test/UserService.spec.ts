import "mockshot";
import { UserService } from "../src/UserService";

describe("UserService", () => {

    // fake database service
    const db = {
        findOneById: async (id) => {
            const name = "Some Name";
            const email = "test@iqoqo.co";
            return {
                id,
                name, 
                email
            } 
        }
    }
    
    it("Should get user", async () => {
        // prepare
        const userService = new UserService(db);
        const id = "abc";

        // execute
        const user = await userService.getUser(id);

        // assert
        expect(user.id).toEqual(id)
        expect(user.name).toBeDefined()
        expect(user.email).toBeDefined()

        expect(user).toMatchMock("UserService", "getUser", "success");

    })
})