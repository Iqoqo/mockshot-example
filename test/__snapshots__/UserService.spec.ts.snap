// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserService Should get user: [mockshot] [ClassSnap] [[UserService getUser success]] 1`] = `
{
  "className": "UserService",
  "methodName": "getUser",
  "mock": {
    "email": "test@iqoqo.co",
    "id": "abc",
    "name": "Some Name"
  },
  "mockName": "success"
}
`;
